### PR TITLE
[FLINK-2861] Fields grouping on split streams fails

### PR DESCRIPTION
**Pull request by mjsax**

### CODE REVIEW

**Comment by mjsax** [comment]
In `FlinkTopologyBuilder` I had to introduce a *token* for operator names which is used in `WrapperSetupHelper` to get the original name of an operator (Flink changes operator names when chaining happens). This is quite hacky. If anybody has a better idea how to solve this, please let me know.

**Comment by StephanEwen** [comment]
I ran into that issue a while back. The problem was that you emit elements that have the name/tag of the split stream (to use the split selector) and over that, you run the regular partitioner.

Splitting this into two steps should work: Splitting, having a mapper that removes the "split tag" and then doing the field partitioning.

Would be nice to make this nice in Flink, but it is tricky. Not all functions go though a collector, and not all collectors can split.

**Comment by mjsax** [comment]
I don't understand you comment? What you describe is exactly how it works (even before this PR). The problem was actually just a missing `TypeInformation` I added here... However, I introduced this `token` into the name of an operator here, because the `.select(...).map(...)` get chained which alters the operator name and I need the original name later on in `WrapperSetupHelper` -- this part is "hacky". But I am not sure if there is a better solution to it. Look for the new variable `FlinkTopologyBuilder.token`.

**Comment by mjsax** [comment]
Btw: Travis fails due to a instable test -- I already create a JIRA for it.

File: flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/wrappers/BoltWrapper.java [hunk 1]
```diff
@@ -209,12 +209,8 @@ public void open() throws Exception {
...
 
 		this.flinkCollector = new TimestampedCollector<OUT>(output);
-		OutputCollector stormCollector = null;
-
-		if (this.numberOfAttributes.size() > 0) {
-			stormCollector = new OutputCollector(new BoltCollector<OUT>(
-					this.numberOfAttributes, flinkCollector));
-		}
+		final OutputCollector stormCollector = new OutputCollector(new BoltCollector<OUT>(
+				this.numberOfAttributes, flinkCollector));
```
Comment (mxm) [inline flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/wrappers/BoltWrapper.java]: Yes, I fixed that in my code base as well...

File: flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/wrappers/WrapperSetupHelper.java [hunk 2]
```diff
@@ -118,6 +119,11 @@ static synchronized TopologyContext createTopologyContext(
 			// prefix "Source: " is inserted by Flink sources by default -- need to get rid of it here
 			operatorName = operatorName.substring(8);
 		}
+		final int idx = operatorName.indexOf(FlinkTopologyBuilder.token);
+		if (idx != -1) {
+			// in case of operator chaining Flink alters the name -- need to get rir of it here
```
Comment (mxm) [inline flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/wrappers/WrapperSetupHelper.java]: Typo: "rid"

**Comment by mxm** [comment]
Maybe a stupid question but why don't you send the operator name along with the BoltWrapper? Then there is no need to extract it from the Flink task name.

**Comment by mjsax** [comment]
That's  actually not a stupid question. I did not do it in the first place to avoid "redundant" code (I was not aware that Flink changes names). I just changed it to your suggestion. If Travis passed, I would merge this. Or are there any objections?

**Comment by StephanEwen** [comment]
@mjsax Okay, I probably misread the original code. I thought the `map()` that drops the tag was not there...

**Comment by mjsax** [comment]
@StephanEwen In the original code that was actually the case. The `BoltWrapper` receiving the data was able to extract the tuples by itself. This got changed as I introduces proper generic types in `FlinkTopologyBuilder`.

Merging this now...

**Comment by StephanEwen** [comment]
Ah, okay, so I skipped a version by accident ;-) I think that original change could not quite have worked, as it would have propagated the wrong TypeInfos to the ArrayKeySelector and OutputSerializer. The one with `map()` looks good...

**Pull request closed without merging**